outlines/generate/generator.py

```diff
@@ -0,0 +1,315 @@
+import dataclasses
+import math
+from typing import TYPE_CHECKING, Callable, List, Optional, Union
+
+import torch
+
+from outlines.index.index import FSMState
+
+if TYPE_CHECKING:
+    from outlines.generate.samplers import Sampler
+    from outlines.index.index import FSM
+
+
+@dataclasses.dataclass(frozen=True)
+class GenerationState:
+    token_ids: torch.Tensor
+    attention_masks: torch.Tensor
+    kv_cache: Optional[torch.Tensor] = None
```

Comment:
Sequence logprob?

Reply:
Also this might be overkill but we may want a separate `Sequence` class that contains the token ids and the logprob.

Reply:
We should handle the KV cache at the model level (which becomes a `Session` object)